pkg/ebpf/c/tracee.bpf.c

```diff
@@ -5008,7 +5015,64 @@ int BPF_KRETPROBE(trace_ret_sock_alloc_file)
     return 0;
 }
 
-static __always_inline u32 security_socket_send_recv_msg(struct socket *sock, event_data_t *event)
+SEC("kprobe/security_sk_clone")
+int BPF_KPROBE(trace_security_sk_clone)
+{
+    // When a "sock" is cloned because of a SYN packet, a new "sock" is created
+    // and the return value is the new "sock" (not the original one).
+    //
+    // There is a problem though, the "sock" does not contain a valid "socket"
+    // associated to it yet (sk_socket is NULL as this is running with SoftIRQ
+    // context). Without a "socket" we also don't have a "file" associated to
+    // it, nor an inode associated to that file. This is the way tracee links
+    // a network flow (packets) to a task.
+    //
+    // The only way we can relate this new "sock", just cloned by a kernel
+    // thread, to a task, is through the existence of the old "sock" struct,
+    // describing the listening socket (one accept() was called for).
+    //
+    // Then, by knowing the old "sock" (with an existing socket, an existing
+    // file, an existing inode), we're able to link this new "sock" to the task
+    // we're tracing for the old "sock".
+    //
+    // In bullets:
+    //
+    // - tracing a process that has a socket listening for connections.
+    // - it receives a SYN packet and a new socket can be created (accept).
+    // - a sock (socket descriptor) is created for the socket to be created.
+    // - no socket/inode exists yet (sock->sk_socket is NULL).
+    // - accept() traces are too late for initial pkts (socked does not exist).
+    // - by linking old "sock" to the new "sock" we can relate the task.
+    // - some of the initial packets, sometimes with big length, are traced now.
+    //
+    // More at: https://github.com/aquasecurity/tracee/issues/2739
+
+    struct sock *osock = (void *) PT_REGS_PARM1(ctx);
```

Comment:
i see that `security_sk_clone` is called from few places in the kernel (and not just for the case of TCP accept). are we interested in updating the map for every family/protocol?
if not, you can check here the family/protocol of the socket.

Reply:
This is what I mapped when I was doing this work:

![image](https://user-images.githubusercontent.com/7395852/232468049-4b4488f1-7f57-432a-93dd-6fce819a42df.png)

I could only find this execution path:

```
sk_set_socket()
  sk_clone_lock() (where security_sk_clone() is called)
    inet_csk_clone_lock()
      tcp_create_openrq_child()
        tcp_v4_sync_recv_sock (also called by tcp_ipv6)
          syn_recv_sock (from ipv4 sock_af_ops)
```

But with a "search" now (instead of a callee function search) I could see it is called by:

- af_alg_accept() (PF_ALG type)
- bluetooth (PF_BLUETOOTH)
- iucv (z/VM comms, PF_IUCV)
- tipc (cluster domain sockets, PF_TIPC)
- vsock (VM / Hypervisor sockets, PF_VSOCK)

So I guess my "kconfig" file (used to index my code) is too small (not including all the other execution paths). So you're right, checking the socket type is likely better (despite those socket types being uncommon or not).